### PR TITLE
Dense Fisher EKF

### PR DESCRIPTION
**Pull request by jcqcai**
I went ahead and gave my shot at implementing dense fisher EKF. Please let me know if anything needs modification. Thank you!

### CODE REVIEW

File: posteriors/ekf/dense_fisher.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    Returns:
+        Updated EKFDenseState.
+    """
+    if not per_sample:
+        log_likelihood = per_samplify(log_likelihood)
+
+    with torch.no_grad(), CatchAuxError():
+        log_liks, aux = log_likelihood(state.params, batch)
+        jac, _ = jacrev(log_likelihood, has_aux=True)(state.params, batch)
+        grad = tree_map(lambda x: x.mean(0), jac)
```
Comment (SamDuffield) [inline posteriors/ekf/dense_fisher.py:154]:
```suggestion
        def log_likelihood_reduced(params, batch):
            per_samp_log_lik, internal_aux = log_likelihood(params, batch)
            return per_samp_log_lik.mean(), internal_aux

        grad, (log_liks, aux) = grad_and_value(log_likelihood_reduced, has_aux=True)(
            state.params, batch
        )
```

Calls to `jacrev` are [significantly more expensive](https://github.com/google/jax/pull/762#issuecomment-1001885437) than `grad`, so we should avoid calling it twice here (there is the internal call to `jacrev` inside `empirical_fisher`).

What do you think of my solution above? (Imports will likely need cleaning up and `from torch.func import grad_and_value` added)

File: tests/ekf/test_dense_fisher.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    # the samples under parameter "b". Reshaping therefore has whole rows of "a" samples before "b" samples begin.
+    # Really, we want each 2D "a" sample followed by a 1D "b" sample, each 3-column row containing a sample of each param in this way.
+    # Therefore, we have to carefully column stack the result of raveling each parameter and reshaping them.
+    # This is all to aid in computing the sample covariance.
+    k = list(samples.keys())[0]  # get first sample to start column stacking
+    samples_copy = tree_ravel(samples[k])[0].reshape(num_samples, samples[k].shape[1])
+    for k in list(samples.keys())[1:]:
+        v = tree_ravel(samples[k])[0].reshape(num_samples, samples[k].shape[1])
+        samples_copy = torch.column_stack((samples_copy, v))
+    samples_cov = torch.cov(samples_copy.T)
```
Comment (SamDuffield) [inline tests/ekf/test_dense_fisher.py:65]:
```suggestion
    flat_samples = torch.vmap(lambda s: tree_ravel(s)[0])(samples)
    samples_cov = torch.cov(flat_samples.T)
```

This seems to provide the same result for me?

**Review by SamDuffield — COMMENTED** [review]
This looks awesome! Only a couple of points to discuss

File: tests/ekf/test_dense_fisher.py (hunk 2 above)
Comment (SamDuffield) [reply tests/ekf/test_dense_fisher.py:65]: Oh! One more thing, we should add a bullet point to [`docs/api/index.md`](https://github.com/normal-computing/posteriors/blob/main/docs/api/index.md?plain=1) briefly describing the algorithm

File: posteriors/ekf/dense_fisher.py (hunk 1 above)
Comment (jcqcai) [reply posteriors/ekf/dense_fisher.py]: Looks good, thanks! Added the import.

File: tests/ekf/test_dense_fisher.py (hunk 2 above)
Comment (jcqcai) [reply tests/ekf/test_dense_fisher.py]: That's much better than what I had. I was naively just calling tree_ravel on the samples directly and then reshaping, which causes the wrong order, so I broke out the for loops. Thanks!

**Review by SamDuffield — APPROVED** [review]
Looks all good to me! Great stuff 🥳

**Pull request merged**